File: 2021/clojure/src/day03.clj
(ns day03
  (:require [clojure.string :as str]
            [utils :refer [parse-integer]]))

(def input
  (->> "data/day03.txt"
       slurp
       str/split-lines))

(defn frequencies-by [xs pred]
  (let [transposed (apply mapv vector xs)
        n (/ (count xs) 2)]
    (mapv (fn [ys]
            (let [freq (->> ys
                            frequencies
                            (sort-by val pred)
                            first)]
              (if (= n (val freq))
                (if (= pred >)
                  \1
                  \0)
                (key freq))))
          transposed)))

(defn gamma-rate [xs]
  (parse-integer 2 (frequencies-by xs >)))

(defn epsilon-rate [xs]
  (parse-integer 2 (frequencies-by xs <)))

(defn generator-rating [xs pred]
  (parse-integer 2
                 (loop [ys xs
                        index 0]
                   (let [n (nth (frequencies-by ys pred) index)
                         zs (filter #(= n (nth % index)) ys)]
                     (if (= 1 (count zs))
                       (first zs)
                       (recur zs (inc index)))))))

(defn oxygen-generator-rating [xs]
  (generator-rating xs >))

(defn co2-generator-rating [xs]
  (generator-rating xs <))

(defn part1 []
  (* (gamma-rate input) (epsilon-rate input)))

(defn part2 []
  (* (oxygen-generator-rating input) (co2-generator-rating input)))

(do
  (println "Part 1:" (part1))
  (println "Part 2:" (part2)))

File: 2021/python/aoc2021/day03.py
from .utils import read_lines_from_file


def part_1(data: list[str]) -> int:
    n = len(data[0])
    gamma = 0
    for i in range(n):
        value = sum(row[i] == "1" for row in data) > len(data) / 2
        gamma += value << (n - 1 - i)
    epsilon = gamma ^ int("1" * n, 2)
    return gamma * epsilon


def part_2(data: list[str]) -> int:
    n = len(data[0])

    oxygen = data.copy()
    for i in range(n):
        value = sum(row[i] == "1" for row in oxygen) >= len(oxygen) / 2
        oxygen = [candidate for candidate in oxygen if int(candidate[i]) == value]
        if len(oxygen) == 1:
            break

    co2 = data.copy()
    for i in range(n):
        value = sum(row[i] == "1" for row in co2) < len(co2) / 2
        co2 = [candidate for candidate in co2 if int(candidate[i]) == value]
        if len(co2) == 1:
            break

    return int(oxygen[0], 2) * int(co2[0], 2)


def main():
    data = read_lines_from_file("day03.txt")
    print("Part 1:", part_1(data))
    print("Part 2:", part_2(data))


if __name__ == "__main__":
    main()
